programs/stake/src/stake_state.rs

```diff
@@ -608,39 +614,25 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
             return Err(InstructionError::InvalidAccountData);
         }
+        if !matches!(split.state()?, StakeState::Uninitialized) {
+            return Err(InstructionError::InvalidAccountData);
+        }
+        if lamports > self.lamports()? {
+            return Err(InstructionError::InsufficientFunds);
+        }
 
-        if let StakeState::Uninitialized = split.state()? {
-            // verify enough account lamports
-            if lamports > self.lamports()? {
-                return Err(InstructionError::InsufficientFunds);
-            }
-
-            match self.state()? {
-                StakeState::Stake(meta, mut stake) => {
-                    meta.authorized.check(signers, StakeAuthorize::Staker)?;
-                    let split_rent_exempt_reserve = calculate_split_rent_exempt_reserve(
-                        meta.rent_exempt_reserve,
-                        self.data_len()? as u64,
-                        split.data_len()? as u64,
-                    );
-
-                    // verify enough lamports for rent and more than 0 stake in new split account
-                    if lamports <= split_rent_exempt_reserve.saturating_sub(split.lamports()?)
-                        // if not full withdrawal
-                        || (lamports != self.lamports()?
-                            // verify more than 0 stake left in previous stake
-                            && checked_add(lamports, meta.rent_exempt_reserve)? >= self.lamports()?)
-                    {
-                        return Err(InstructionError::InsufficientFunds);
-                    }
-                    // split the stake, subtract rent_exempt_balance unless
-                    // the destination account already has those lamports
-                    // in place.
-                    // this means that the new stake account will have a stake equivalent to
-                    // lamports minus rent_exempt_reserve if it starts out with a zero balance
-                    let (remaining_stake_delta, split_stake_amount) = if lamports
-                        == self.lamports()?
-                    {
+        match self.state()? {
+            StakeState::Stake(meta, mut stake) => {
+                meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                let validated_split_info = validate_split_amount(self, split, lamports, &meta)?;
+
+                // split the stake, subtract rent_exempt_balance unless
+                // the destination account already has those lamports
+                // in place.
+                // this means that the new stake account will have a stake equivalent to
+                // lamports minus rent_exempt_reserve if it starts out with a zero balance
+                let (remaining_stake_delta, split_stake_amount) =
+                    if validated_split_info.source_remaining_balance == 0 {
                         // If split amount equals the full source stake, the new split stake must
```

Comment:
nit: this comment could be tweaked to avoid being stale a bit: `If split amount equals the full source stake as implied by 0 source_remaining_balance, the new split stake ...`.

Reply:
Done!